File: 572_subtree_of_another_tree/572_subtree_of_another_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
var isSubtree = function (root, subRoot) {
  // function serialize(node) {
  //     if (!node) {
  //         return "null";
  //     }
  //     return "#" + node.val + serialize(node.left) + serialize(node.right);
  // }

  // const rootStr = serialize(root);
  // const subRootStr = serialize(subRoot);

  // return rootStr.includes(subRootStr);
  function isSameTree(tree1, tree2) {
    if (!tree1 && !tree2) {
      return true;
    }
    if (!tree1 || !tree2 || tree1.val !== tree2.val) {
      return false;
    }
    return (
      isSameTree(tree1.left, tree2.left) && isSameTree(tree1.right, tree2.right)
    );
  }

  // 如果當前節點相同，則返回 true
  if (!root) {
    return false;
  }
  if (isSameTree(root, subRoot)) {
    return true;
  }

  // 否則，遞歸檢查左子樹和右子樹
  return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
};

File: 572_subtree_of_another_tree/572_subtree_of_another_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root:  # If root is None, return False
            return False
        # Check if the current subtree starting at root matches subRoot
        return self.isSameTree(root, subRoot) or \
               self.isSubtree(root.left, subRoot) or \
               self.isSubtree(root.right, subRoot)

    def isSameTree(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        if not root1 and not root2:  # Both trees are empty, hence same
            return True
        if not root1 or not root2:  # One of the trees is empty, not same
            return False
        # Check if current nodes have the same value and if subtrees are same
        return root1.val == root2.val and \
               self.isSameTree(root1.left, root2.left) and \
               self.isSameTree(root1.right, root2.right)
